// moralis/lib/browser/MoralisDot.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _slice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/slice"));

var _from = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/array/from"));

var _indexOf = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/index-of"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _ParseUser = _interopRequireDefault(require("./ParseUser"));

var _ParseQuery = _interopRequireDefault(require("./ParseQuery"));

var _ParseObject = _interopRequireDefault(require("./ParseObject"));

var _ParseACL = _interopRequireDefault(require("./ParseACL")); // import {
//   web3Accounts,
//   web3Enable,
//   web3FromAddress,
// } from '@polkadot/extension-dapp';
// import { stringToU8a } from '@polkadot/util/string/toU8a'


var web3EnablePromise = null;

var MoralisDot = /*#__PURE__*/function () {
  function MoralisDot() {
    (0, _classCallCheck2.default)(this, MoralisDot);
  }

  (0, _createClass2.default)(MoralisDot, null, [{
    key: "web3IsInjected",
    value: function () {
      return (0, _keys.default)(window.injectedWeb3).length !== 0;
    }
  }, {
    key: "enable",
    value: function () {
      var _enable = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(opts) {
        var _window$injectedWeb, _window$injectedWeb$t;

        var type,
            _args = arguments;
        return _regenerator.default.wrap(function (_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                type = _args.length > 1 && _args[1] !== undefined ? _args[1] : 'polkadot-js';

                if (!web3EnablePromise) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt("return", web3EnablePromise);

              case 3:
                web3EnablePromise = (_window$injectedWeb = window.injectedWeb3) === null || _window$injectedWeb === void 0 ? void 0 : (_window$injectedWeb$t = _window$injectedWeb[type]) === null || _window$injectedWeb$t === void 0 ? void 0 : _window$injectedWeb$t.enable(opts);
                return _context.abrupt("return", web3EnablePromise);

              case 5:
              case "end":
                return _context.stop();
            }
          }
        }, _callee);
      }));

      return function () {
        return _enable.apply(this, arguments);
      };
    }()
  }, {
    key: "authenticate",
    value: function () {
      var _authenticate = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(opts) {
        var _opts$name, _context2, _user$get;

        var allAccounts, account, address, dotAddress, accounts, data, signature, authData, user;
        return _regenerator.default.wrap(function (_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                _context3.next = 2;
                return MoralisDot.enable((_opts$name = opts === null || opts === void 0 ? void 0 : opts.name) !== null && _opts$name !== void 0 ? _opts$name : 'Moralis');

              case 2:
                MoralisDot.web3 = _context3.sent;
                _context3.next = 5;
                return MoralisDot.web3.accounts.get();

              case 5:
                allAccounts = _context3.sent;
                account = allAccounts[0];
                address = account === null || account === void 0 ? void 0 : account.address;

                if (address) {
                  _context3.next = 10;
                  break;
                }

                throw new Error('Address not found');

              case 10:
                // const account = await proxy.getAccount(address);
                dotAddress = address; //.toLowerCase();

                accounts = [dotAddress];
                data = MoralisDot.getSigningData();
                _context3.next = 15;
                return MoralisDot.sign(address, data);

              case 15:
                signature = _context3.sent;
                authData = {
                  id: dotAddress,
                  signature: signature,
                  data: data
                };
                console.log(authData);
                _context3.next = 20;
                return _ParseUser.default.logInWith('moralisDot', {
                  authData: authData
                });

              case 20:
                user = _context3.sent;
                _context3.next = 23;
                return user.setACL(new _ParseACL.default(user));

              case 23:
                if (user) {
                  _context3.next = 25;
                  break;
                }

                throw new Error('Could not get user');

              case 25:
                user.set('dotAccounts', uniq((0, _concat.default)(_context2 = []).call(_context2, accounts, (_user$get = user.get('dotAccounts')) !== null && _user$get !== void 0 ? _user$get : [])));
                user.set('dotAddress', dotAddress);
                _context3.next = 29;
                return user.save();

              case 29:
                return _context3.abrupt("return", user);

              case 30:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee2);
      }));

      return function () {
        return _authenticate.apply(this, arguments);
      };
    }()
  }, {
    key: "link",
    value: function () {
      var _link = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3(account) {
        var _context4, _user$get2;

        var user, dotAddress, DotAddress, query, dotAddressRecord, data, signature, authData;
        return _regenerator.default.wrap(function (_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                _context5.next = 2;
                return _ParseUser.default.current();

              case 2:
                user = _context5.sent;
                dotAddress = account; //.toLowerCase();

                DotAddress = _ParseObject.default.extend('_DotAddress');
                query = new _ParseQuery.default(DotAddress);
                _context5.next = 8;
                return query.get(dotAddress).catch(function () {
                  return null;
                });

              case 8:
                dotAddressRecord = _context5.sent;

                if (dotAddressRecord) {
                  _context5.next = 17;
                  break;
                }

                data = MoralisDot.getSigningData();
                _context5.next = 13;
                return MoralisDot.sign(dotAddress, data);

              case 13:
                signature = _context5.sent;
                authData = {
                  id: dotAddress,
                  signature: signature,
                  data: data
                };
                _context5.next = 17;
                return user.linkWith('moralisDot', {
                  authData: authData
                });

              case 17:
                user.set('dotAccounts', uniq((0, _concat.default)(_context4 = [dotAddress]).call(_context4, (_user$get2 = user.get('dotAccounts')) !== null && _user$get2 !== void 0 ? _user$get2 : [])));
                user.set('dotAddress', dotAddress);
                _context5.next = 21;
                return user.save();

              case 21:
                return _context5.abrupt("return", user);

              case 22:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee3);
      }));

      return function () {
        return _link.apply(this, arguments);
      };
    }()
  }, {
    key: "unlink",
    value: function () {
      var _unlink = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4(account) {
        var _user$get3;

        var accountsLower, DotAddress, query, dotAddressRecord, user, accounts, nextAccounts;
        return _regenerator.default.wrap(function (_context6) {
          while (1) {
            switch (_context6.prev = _context6.next) {
              case 0:
                accountsLower = account; //.toLowerCase();

                DotAddress = _ParseObject.default.extend('_DotAddress');
                query = new _ParseQuery.default(DotAddress);
                _context6.next = 5;
                return query.get(accountsLower);

              case 5:
                dotAddressRecord = _context6.sent;
                _context6.next = 8;
                return dotAddressRecord.destroy();

              case 8:
                _context6.next = 10;
                return _ParseUser.default.current();

              case 10:
                user = _context6.sent;
                accounts = (_user$get3 = user.get('dotAccounts')) !== null && _user$get3 !== void 0 ? _user$get3 : [];
                nextAccounts = (0, _filter.default)(accounts).call(accounts, function (v) {
                  return v !== accountsLower;
                });
                user.set('dotAccounts', nextAccounts);
                user.set('dotAddress', nextAccounts[0]);
                _context6.next = 17;
                return user._unlinkFrom('moralisDot');

              case 17:
                _context6.next = 19;
                return user.save();

              case 19:
                return _context6.abrupt("return", user);

              case 20:
              case "end":
                return _context6.stop();
            }
          }
        }, _callee4);
      }));

      return function () {
        return _unlink.apply(this, arguments);
      };
    }()
  }, {
    key: "sign",
    value: function () {
      var _sign = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee5(address, data) {
        var _web3$signer;

        var web3, _yield$web3$signer$si, signature;

        return _regenerator.default.wrap(function (_context7) {
          while (1) {
            switch (_context7.prev = _context7.next) {
              case 0:
                if (web3EnablePromise) {
                  _context7.next = 2;
                  break;
                }

                throw new Error('Must enable MoralisDot');

              case 2:
                _context7.next = 4;
                return web3EnablePromise;

              case 4:
                web3 = _context7.sent;
                _context7.next = 7;
                return (_web3$signer = web3.signer) === null || _web3$signer === void 0 ? void 0 : _web3$signer.signRaw({
                  address: address,
                  data: stringToHex(data),
                  type: 'bytes'
                });

              case 7:
                _yield$web3$signer$si = _context7.sent;
                signature = _yield$web3$signer$si.signature;
                return _context7.abrupt("return", signature);

              case 10:
              case "end":
                return _context7.stop();
            }
          }
        }, _callee5);
      }));

      return function () {
        return _sign.apply(this, arguments);
      };
    }()
  }, {
    key: "getSigningData",
    value: function () {
      return 'Moralis Authentication';
    }
  }]);
  return MoralisDot;
}();

function uniq(arr) {
  return (0, _filter.default)(arr).call(arr, function (v, i) {
    return (0, _indexOf.default)(arr).call(arr, v) === i;
  });
}

var _default = MoralisDot;
exports.default = _default;

function stringToHex(value) {
  return toHexString(stringToU8a(value));
}

function stringToU8a(value) {
  var u8a = new Uint8Array(value.length);

  for (var i = 0; i < value.length; i++) {
    u8a[i] = value.charCodeAt(i);
  }

  return u8a;
}

function toHexString(byteArray) {
  return '0x' + (0, _from.default)(byteArray, function (byte) {
    var _context8;

    return (0, _slice.default)(_context8 = '0' + (byte & 0xFF).toString(16)).call(_context8, -2);
  }).join('');
}
